Cython/Compiler/ExprNodes.py

```diff
@@ -14271,8 +14271,34 @@ def analyse_type_annotation(self, env, assigned_value=None):
         with env.new_c_type_context(in_c_type_context=explicit_ctype):
             arg_type = annotation.analyse_as_type(env)
 
+
         if arg_type is None:
-            warning(annotation.pos, "Unknown type declaration in annotation, ignoring")
+            if isinstance(annotation, NameNode):
+                # Validate annotation in form `var: type`
+                if not env.lookup(annotation.name):
+                    warning(annotation.pos, "Unknown type declaration '%s' in annotation, ignoring" % annotation.name, level=1)
+            elif isinstance(annotation, AttributeNode):
+                # Validate annotation in form `var: module.type`
+                if not env.lookup(annotation.obj.name):
+                    # `module` is undeclared
+                    warning(annotation.pos,
+                            "Unknown type declaration '%s.%s' in annotation, ignoring" % (annotation.obj.name, annotation.attribute),
+                            level=1)
+                elif annotation.obj.is_cython_module:
+                    # `module` is cython
+                    module_scope = annotation.obj.analyse_as_module(env)
+                    if module_scope and not module_scope.lookup_type(annotation.attribute):
+                        warning(annotation.pos,
+                                "Unknown type declaration '%s.%s' in annotation, ignoring" % (annotation.obj.name, annotation.attribute),
+                                level=1)
+                else:
+                    module_scope = annotation.obj.analyse_as_module(env)
+                    if module_scope and module_scope.pxd_file_loaded and not annotation.analyse_as_type(env):
```

Comment:
We already called `annotation.analyse_as_type()` above, that failed and brought us here.

Reply:
fixed.